Mesh 3 : initial_points_generator parameter for make_mesh_3 point initialization (#7798)
CGAL/cgal: Mesh_3/doc/Mesh_3/Mesh_3.txt

\subsubsection Mesh_3DomainsFrom3DImagesWithCustomInitialization Domains From 3D Images, with a Custom Initialization

The example \ref Mesh_3/mesh_3D_image_with_image_initialization.cpp is a modification
of \ref Mesh_3/mesh_3D_image.cpp. The goal of that example is to show how
the default initialization of the triangulation, using random rays, can be
replaced by a new implementation. In this case, the initialization detects
all connected components in the 3D segmented image, and inserts points in
the triangulation for each connected component.

\snippet Mesh_3/mesh_3D_image_with_image_initialization.cpp Meshing

The parameter `CGAL::parameters::initial_points_generator` is used.
It expects a functor that returns a set of points for the mesh

> initialization (following the `InitialPointsGenerator`). The functor

Reviewer: ```suggestion
initialization (following the concept `InitialPointsGenerator`). The functor
```

`CGAL/Mesh_3/Construct_initial_points_labeled_image.h` is used in this example.
It constructs points using the API of the mesh domain, as follows.

> First the functor `construct_intersect` is created

Reviewer 2: ```suggestion
First the functor `construct_intersection` is created
```

> First the functor `construct_intersection` is created

Reviewer: ```suggestion
First, the functor `construct_intersection` is created:
```

the default initialization of the triangulation, using random rays, can be
replaced by a new implementation. In this case, the initialization detects
all connected components in the 3D segmented image, and inserts points in
the triangulation for each connected component.

\snippet Mesh_3/mesh_3D_image_with_image_initialization.cpp Meshing

The parameter `CGAL::parameters::initial_points_generator` is used.
It expects a functor that returns a set of points for the mesh
initialization (following the concept `InitialPointsGenerator_3`). The functor
`Construct_initial_points_labeled_image` is used in this example.
It constructs points using the API of the mesh domain, as follows.
First, the functor `construct_intersection` is created:

\snippet CGAL/Mesh_3/Construct_initial_points_labeled_image.h construct intersection
then the `%Mesh_domain::Intersection` object (a `%tuple` with three
elements) is constructed using a call to the functor `construct_intersection`
\snippet CGAL/Mesh_3/Construct_initial_points_labeled_image.h use construct intersection
and eventually `%intersect_index` is the element \#1 of `%intersect`.

> \snippet CGAL/Mesh_3/Construct_initial_points_labeled_image.h get construct intersection.

Reviewer 3: Overnight documentation build gave:
```
/home/cgal-testsuite/cgal_doc_build/CGAL-6.1-Ic-66/doc/Mesh_3/Mesh_3.txt:761: warning: block marked with [get construct intersection.] for \snippet should appear twice in file CGAL/Mesh_3/Construct_initial_points_labeled_image.h, found it 0 times
```

Just drop the `.` at the end.

Reviewer 4: @janetournois So Doxygen does show errors, when the snippets are not used correctly. 😌

\snippet Mesh_3/mesh_3D_image_with_custom_initialization.cpp Meshing

The code of the function `initialize_triangulation_from_labeled_image()` is
in the header \ref

> CGAL/Mesh_3/initialize_triangulation_from_labeled_image.h. It get points from

Reviewer 2: ```suggestion
CGAL/Mesh_3/initialize_triangulation_from_labeled_image.h. It gets points from
```

> \ref CGAL/Mesh_3/Construct_initial_points_labeled_image.h and insert them

Reviewer 2: ```suggestion
\ref CGAL/Mesh_3/Construct_initial_points_labeled_image.h and inserts them
```

> in the triangulation. The code of the Construct_initial_points_labeled_image

Reviewer 2: ```suggestion
in the triangulation. The code of the `Construct_initial_points_labeled_image`
```

> function is rather complicated. The following lines show how to insert new

Reviewer 2: ```suggestion
functor is rather complicated. The following lines show how to insert new
```

Right: the mesh generated after the initialization of all connected components
\cgalFigureCaptionEnd

Note that the example \ref
Mesh_3/mesh_3D_image_with_initial_points.cpp also shows how to
create a 3D image using the undocumented API of CGAL_ImageIO.

\snippet Mesh_3/mesh_3D_image_with_initial_points.cpp Create the image

The code of the function `%random_labeled_image()` is in the header file \ref
Mesh_3/random_labeled_image.h.

The example \ref Mesh_3/mesh_3D_image_with_custom_initialization.cpp features

> a custom functor that initialize the triangulation.

Reviewer 2: ```suggestion
a custom functor that initializes the triangulation.
```

Right: the mesh generated after the initialization of all connected components
\cgalFigureCaptionEnd

Note that the example \ref
Mesh_3/mesh_3D_image_with_image_initialization.cpp also shows how to
create a 3D image using the undocumented API of CGAL_ImageIO.

> \snippet Mesh_3/mesh_3D_image_with_image_initialization.cpp Create the image

Reviewer 2: this is fixed

\snippet Mesh_3/mesh_3D_gray_image_with_custom_initialization.cpp Meshing

The code of the function `initialize_triangulation_from_gray_image()` is
in the header \ref
CGAL/Mesh_3/initialize_triangulation_from_gray_image.h.

The example \ref Mesh_3/mesh_3D_image_with_initial_points.cpp is another
way to achieve the same results. Instead of a custom initialization,
it uses the parameter `CGAL::parameters::initial_points_generator` for the function

> `CGAL::make_mesh_3`. This parameter expect a functor that ouputs points for

Reviewer 2: ```suggestion
`CGAL::make_mesh_3`. This parameter expects a functor that returns a set of points for
```
